File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <functional>
#include <iostream>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int d = triangle.back().size();
    int dirx[2] = {-1,-1};
    int diry[2] = {0,-1};
    
    vector<vector<int>> dp(d, vector<int>(d,-1));
    dp[0][0] = triangle[0][0];
    dp[1][0] = dp[0][0] + triangle[1][0];
    dp[1][1] = dp[0][0] + triangle[1][1];
    
    int x = 2;
    while (x < d) {
        for(int y=0;y<=x;y++) {
            int now = triangle[x][y];
            int prev_max = 0;
            for(int d=0;d<2;d++) {
                int prevx = x + dirx[d];
                int prevy = y + diry[d];
                if(0<=prevy && prevy < x) {
                    prev_max = max(prev_max, dp[prevx][prevy]);
                } 
            }
            dp[x][y] = now + prev_max;
        }
        x++;
    }
    
    int answer = *max_element(dp.back().begin(), dp.back().end());
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    public int solution(int[][] triangle) {
        int h = triangle.length;
        
        int[][] dp = new int[h][h];
        dp[0][0] = triangle[0][0];
        
        int height = 2;
        while (height <= h) {
            for(int i=0;i<height;i++) {
                if (i == 0) { // 첫번째값
                    dp[height-1][i] = triangle[height-1][i] + dp[height-2][i];
                } else if (i == height-1) { // 끝값
                    dp[height-1][i] = triangle[height-1][i] + dp[height-2][i-1];
                } else { // 중간값
                    dp[height-1][i] = triangle[height-1][i] + Math.max(dp[height-2][i-1], dp[height-2][i]);
                }
            }
            height++;
        }
        
        int answer = 0;
        for(int x: dp[h-1]) {
            answer = Math.max(answer, x);
        }
        
        return answer;
    }
}